lib/slack/web/faraday/response/raise_error.rb

```diff
@@ -15,6 +15,14 @@ def on_complete(env)
             error_class ||= Slack::Web::Api::Errors::SlackError
             raise error_class.new(error_message, env.response)
           end
+
+          def call(env)
+            super
+          rescue Slack::Web::Api::Errors::SlackError, Slack::Web::Api::Errors::TooManyRequestsError => e
+            raise e
+          rescue ::Faraday::ServerError, ::Faraday::ParsingError
+            raise Slack::Web::Api::Errors::InternalError.new('internal_error', env.response)
```

Comment:
I think you want to do the inverse here, handle only errors that are predictable. We don't want to rescue all faraday errors and narrow them down into an `InternalError`, but we do want to rescue predictable situations, such as the server had an internal error and spit out a bunch of non-JSON into a specific `Slack::Web::Api::Errors::InternalError`. The latter should have an inner exception that gets assigned `e`, along with `env.response`, and the message should be one from `::Faraday::ParsingError` I imagine.

Reply:
From the testcase in spec:
```
[2] pry(#<RSpec::ExampleGroups::SlackWebFaradayResponseRaiseError::Call::WithNonJsonReply>)> _ex_
=> #<Slack::Web::Api::Errors::InternalError #<Slack::Web::Api::Errors::InternalError: internal_error>>
[3] pry(#<RSpec::ExampleGroups::SlackWebFaradayResponseRaiseError::Call::WithNonJsonReply>)> _ex_.cause
=> #<Faraday::ParsingError #<Faraday::ParsingError: oops>>
```
So we have Faraday error in `e.cause` and can do anything with it anyway. 

Could you please elaborate on `We don't want to rescue all faraday errors and narrow them down into an InternalError, but we do want to rescue predictable situations, such as the server had an internal error and spit out a bunch of non-JSON into a specific Slack::Web::Api::Errors::InternalError.`?
Should we subclass `Slack::Web::Api::Errors::InternalError` into multiple ones?

Reply:
You're rescuing any faraday error that has specifics to it (e.g. timed out, parsing failed, etc.) into some generic internal error. That seems like it's providing less value than before (narrows it down). Unless I'm misunderstanding something.

Reply:
Well, we cannot predict what errors slack will fails with, but `::Faraday::ServerError` is basically 5xx & timeouts (which we should handle) and `::Faraday::ParsingError` is non-json (which we also should handle). So I don't think that we.re overrescuing here.

Faraday exception is always available via `e.cause`, source exception (i. e. `Net::ReadTimeout`) is available via `e.cause.cause`, thus even while we're narrowing info in the exception itself, no info is lost and everything is accessible if needed.

We can do something like (naming i hard)
```ruby
module Slack::Web::Api::Errors
  class ServerError < InternalError; end # common class
  class ParsingError < Server; end # wrap ::Faraday::ParsingError
  class ConnectionError < InternalError; end # just commong class
  class TimeoutError < ConnectionError; end # For ::Faraday::ServerError where cause is timeout
  class UnavailableError < ConnectionError; end # For ::Faraday::ServerError where status is 5xx
end
```
if you wish & raise different exceptions for different causes, but still be able to catch them via `rescue Slack::Web::Api::Errors::Error` or more precise ones.

Reply:
I like this, but I am not sure what the right answer is here, put up a PR that you think should/is ready to be merged and I'll give it more thought?

Reply:
Pushed, not sure about naming/messages in errors and there is behaviour change: we're raising `SlackError` instead of `Faraday::Error` if server is unavailable, I can revert that one if you wish (and fix the typo in the disabled spec).